src/storage/mod.rs

```diff
@@ -976,6 +976,179 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
+                ctx.get_request_source(),
+                ctx.get_resource_control_context().get_override_priority(),
+            )
+        });
+        let priority_tag = get_priority_tag(priority);
+        let key_ranges = keys
+            .iter()
+            .map(|k| (k.as_encoded().to_vec(), k.as_encoded().to_vec()))
+            .collect();
+        let resource_tag = self
+            .resource_tag_factory
+            .new_tag_with_key_ranges(&ctx, key_ranges);
+        let concurrency_manager = self.concurrency_manager.clone();
+        let api_version = self.api_version;
+        let busy_threshold = Duration::from_millis(ctx.busy_threshold_ms as u64);
+        let quota_limiter = self.quota_limiter.clone();
+        let mut sample = quota_limiter.new_sample(true);
+        self.read_pool_spawn_with_busy_check(
+            busy_threshold,
+            async move {
+                let stage_scheduled_ts = Instant::now();
+                let mut key_ranges = vec![];
+                for key in &keys {
+                    key_ranges.push(build_key_range(key.as_encoded(), key.as_encoded(), false));
+                }
+                tls_collect_query_batch(
+                    ctx.get_region_id(),
+                    ctx.get_peer(),
+                    key_ranges,
+                    QueryKind::Get,
+                );
+
+                KV_COMMAND_COUNTER_VEC_STATIC.get(CMD).inc();
+                SCHED_COMMANDS_PRI_COUNTER_VEC_STATIC
+                    .get(priority_tag)
+                    .inc();
+
+                deadline.check()?;
+
+                Self::check_api_version(
+                    api_version,
+                    ctx.api_version,
+                    CMD,
+                    keys.iter().map(Key::as_encoded),
+                )?;
+
+                let command_duration = Instant::now();
+
+                let snap_ctx = prepare_snap_ctx(
+                    &ctx,
+                    &keys,
+                    start_ts,
+                    // buffer_batch_get doesn't read locks from other txns
+                    &TsSet::Empty,
+                    &concurrency_manager,
+                    CMD,
+                )?;
+                let snapshot =
+                    Self::with_tls_engine(|engine| Self::snapshot(engine, snap_ctx)).await?;
+                {
+                    deadline.check()?;
+                    let begin_instant = Instant::now();
+
+                    let stage_snap_recv_ts = begin_instant;
+                    let (result, stats) = Self::with_perf_context(CMD, || {
+                        let _guard = sample.observe_cpu();
+                        let mut reader = MvccReader::new(
+                            snapshot,
+                            Some(ScanMode::Forward),
+                            !ctx.get_not_fill_cache(),
+                        );
+                        // TODO: metrics
+                        // TODO: refactor: reuse functions in PointGetter
```

Comment:
If the command is planned to support `UnionStoreBatchGet` in the future, another
``` 
TODO:  Support merging results of membuf read and snapshot read.
```
is needed

Reply:
I don't think it necessary because current code doesn't show any indication of a need of UnionGet. And we have not yet decided the interface of such a union get.
I put TODOs for stats and refactoring, as other readers might wonder about these aspects.